lib/Dialect/ESI/capnp/CPPAPI.cpp

```diff
@@ -0,0 +1,449 @@
+
+#include "ESICapnp.h"
+#include "circt/Dialect/Comb/CombOps.h"
+#include "circt/Dialect/ESI/ESITypes.h"
+#include "circt/Dialect/HW/HWDialect.h"
+#include "circt/Dialect/HW/HWOps.h"
+#include "circt/Dialect/HW/HWTypes.h"
+#include "circt/Dialect/SV/SVOps.h"
+#include "circt/Support/IndentingOStream.h"
+
+#include "capnp/schema-parser.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/IR/Builders.h"
+#include "mlir/IR/BuiltinAttributes.h"
+#include "mlir/IR/BuiltinTypes.h"
+#include "llvm/ADT/IntervalMap.h"
+#include "llvm/ADT/TypeSwitch.h"
+#include "llvm/Support/Format.h"
+
+#include <initializer_list>
+#include <string>
+
+using namespace circt::esi::capnp::detail;
+using namespace circt;
+using namespace support;
+
+//===----------------------------------------------------------------------===//
+// CPPType class implementation.
+//===----------------------------------------------------------------------===//
+
+static std::string lowercase(llvm::StringRef str) {
+  std::string s;
+  llvm::raw_string_ostream ss(s);
+  for (auto &c : str)
+    ss << llvm::toLower(c);
+  return s;
+}
+
+static std::string joinStringAttrArray(mlir::ArrayAttr strings,
+                                       llvm::StringRef delimiter) {
+  std::string s;
+  llvm::raw_string_ostream ss(s);
+  auto stringRefs = SmallVector<StringRef>(
+      llvm::to_vector(strings.getAsValueRange<StringAttr>()));
+  llvm::interleave(
+      stringRefs, ss, [&](auto str) { ss << str; }, delimiter);
+  return s;
+}
+
+static void emitCPPType(Type type, llvm::raw_ostream &os) {
+  llvm::TypeSwitch<Type>(type)
+      .Case([&os](IntegerType intTy) {
+        auto w = intTy.getWidth();
+        if (w == 0) {
+          os << "void";
+        } else if (w == 1) {
+          os << "bool";
+        } else {
+          if (intTy.isSigned())
+            os << "int";
+          else
+            os << "uint";
+
+          // Round up.
+          if (w <= 8)
+            os << "8";
+          else if (w <= 16)
+            os << "16";
+          else if (w <= 32)
+            os << "32";
+          else if (w <= 64)
+            os << "64";
+          else
+            assert(false && "Type not supported. Integer too wide. Please "
+                            "check support first with isSupported()");
+
+          os << "_t";
+        }
+      })
+      .Case<hw::ArrayType, hw::StructType>([](auto Ty) {
+        assert(false && "Structs containing List and Structs not supported");
+      })
+      .Default([](Type) {
+        assert(false && "Type not supported. Please check support first with "
+                        "isSupported()");
+      });
+}
+
+circt::esi::capnp::CPPType::CPPType(mlir::Type type) : ESICapnpType(type) {}
+
+llvm::StringRef circt::esi::capnp::CPPType::cppName() const {
+  return capnpName();
+}
+
+static bool isZeroWidthInt(Type type) {
+  return type.isa<IntegerType>() && type.cast<IntegerType>().getWidth() == 0;
```

Comment:
This should also include `NoneType`, correct?

Reply:
correct!